### ISSUE TITLE
Mozilla test ecma/TypeConversion/9.2.js fails in debug mode

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#28]
This test fails on the simulator, and I am pretty sure also on hardware, though it is worth re-testing that.

Here is a full command:

```
out/mipsel.debug/d8 --stress-opt --always-opt --nobreak-on-abort --nodead-code-elimination --enable-slow-asserts --debug-code --verify-heap --expose-gc test/mozilla/mozilla-shell-emulation.js --es5_readonly test/mozilla/data/shell.js test/mozilla/data/ecma/shell.js test/mozilla/data/ecma/TypeConversion/shell.js test/mozilla/data/ecma/TypeConversion/9.2.js --test
```

I verified that this test does _not_ use Long Branches, so my discussion of that in group call was erroneous. 

Here is a view of the output so you can see the one failing subtest, which happens only in stress-opt/always-opt, in stress 2/2:

```
PASSED! NaN ? true : false = false expected: false
PASSED! 1000 % 0 ? true : false = false expected: false
FAILED! (function(a,b){ return a % b ? true : false })(1000, 0) = true expected: false
PASSED! (function(x) { return !(x) })(0/0) = true expected: true
PASSED! !(0/0) = true expected: true
```

can you please debug this, Dusan?

**Comment by dusmil** [comment]
It appears this bug is critical as it is present in mips code and can be reproduced in release mode. It has remained hidden since our initial implementations of crankshaft.
We are lucky the bug occurs in rare edge case that is triggered only during stress testing.
Because of lucky circumstances of specific testing sequence of debug mode it is exposed only in that mode at first.

I will try to explain why it can be seen in debug mode by describing stress testing sequence in debug and release mode.

When --stress-opt flags is passed  for debug mode test is run twice  and for release five times with different flags to d8.

Debug mode:
1. pass 
   d8 run with lazy Optimizations  
2. pass
   d8 run with forced optimizations

Release mode;
1. pass
   d8 run with lazy Optimizations  
2. pass
   d8 run with lazy Optimizations  
3. pass
   d8 run with lazy Optimizations  
4. pass
   d8 normal v8 operational mode ( no flags passed)
5. pass
   d8 run with forced optimizations

Forced optimizations means that every code is compiled with crankshaft no matter what ( Even runtime  type collections is not done, so  default assumptions about type internal representations are used).

Lazy optimization means function is initially compiled with full compiler, and in next invocation it is compiled with crankshaft.

Normal mode  is ordinary mode where everything is initially compiled with full compiler and  profiler decides when to compile function (there is enough type feedback info)

The bug occurs only when forced optimization is performed immediately after lazy optimizations as it is debug mode. If that test sequence is used in release mode test will fail. This can be achieved by altering functions GetStressRuns and PrepareStressRun in api.cc.

Then at that  moment  internal number representation  is switched to double in optimized code.

Bottom line, our implementation of DoBranch and BranchF is faulty, and does not consider Nan value. The consequence of this is that Nan value in boolean condition is not interpreted as false.

The fix for this is on dm-dev branch.

**Comment by paul99** [comment]
This fix was landed at google today.

SHA e3b3f07bb00312832ac9a40c684ea0f85faf8726

MIPS: Fixed faulty branch condition handling for doubles.

r 13849

Closing issue.
